packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {HttpPlugin} from '@opencensus/instrumentation-http';
+import {types} from '@opencensus/opencensus-core';
+import {classes} from '@opencensus/opencensus-core';
+import {logger} from '@opencensus/opencensus-core';
+import * as http from 'http';
+import * as http2 from 'http2';
+import * as net from 'net';
+import * as shimmer from 'shimmer';
+import * as tls from 'tls';
+import * as url from 'url';
+import * as uuid from 'uuid';
+
+export type Http2Module = typeof http2;
+export type ConnectFunction = typeof http2.connect;
+type RequestFunction =
+    (this: http2.ClientHttp2Session, headers?: http2.OutgoingHttpHeaders,
+     options?: http2.ClientSessionRequestOptions) => http2.ClientHttp2Stream;
+export type CreateServerFunction = typeof http2.createServer;
+
+/** Http2 instrumentation plugin for Opencensus */
+export class Http2Plugin extends HttpPlugin {
+  server: http2.Http2Server|http2.Http2SecureServer;
+  client: http2.ClientHttp2Session;
+  clientUrl: url.UrlWithStringQuery;
+  patchedCreateServer = false;
+  patchedConnect = false;
+
+  /** Constructs a new Http2Plugin instance. */
+  constructor() {
+    super('http2');
+  }
+
+  /**
+   * Patches HTTP2 incoming and outcoming request functions.
+   * @param moduleExporters The HTTP2 package.
+   * @param tracer A tracer instance to create spans on.
+   * @param version The package version.
+   */
+  // tslint:disable-next-line:no-any
+  applyPatch(moduleExports: any, tracer: types.Tracer, version: string) {
+    this.setPluginContext(moduleExports, tracer, version);
+    this.logger = tracer.logger || logger.logger('debug');
+
+    shimmer.wrap(moduleExports, 'createServer', this.patchCreateServer());
+    shimmer.wrap(moduleExports, 'createSecureServer', this.patchCreateServer());
+
+    shimmer.wrap(moduleExports, 'connect', this.patchConnect());
+
+    return moduleExports;
+  }
+
+  /** Unpatches all HTTP2 patched function. */
+  applyUnpatch(): void {
+    if (this.patchedCreateServer) {
+      shimmer.unwrap(plugin.server.constructor.prototype, 'emit');
+    } else {
+      shimmer.unwrap(plugin.moduleExports, 'createServer');
+      shimmer.unwrap(plugin.moduleExports, 'createSecureServer');
+    }
+
+    if (this.patchedConnect) {
+      shimmer.unwrap(plugin.client, 'request');
+    } else {
+      shimmer.unwrap(plugin.moduleExports, 'connect');
+    }
+  }
+
+  patchConnect() {
```

Comment:
These functions should be renamed. They don't patch the corresponding functions, but rather return a function that patches the corresponding functions. A name like `getPatchConnectFn` would be more appropriate.

Reply:
Makes total sense. We choose this name because the inner function is anonymous and we didn't see a problem on calling it this way. Also, HTTP uses this "name pattern", if you think it's best to change it here, I suggest we change it on HTTP as well.

Reply:
Yep, I would be in favor of that.